improver_tests/wxcode/test_WeatherSymbols.py

```diff
@@ -0,0 +1,607 @@
+) -> None:
+    """Runs the WeatherSymbols plugin with the supplied inputs and asserts that the resulting
+    weather code matches the expected symbol
+
+    Args:
+        expected:
+            Weather symbol meaning that we expect the WeatherSymbols plugin to produce
+        wxcode_inputs:
+            All the input cubes to give to the plugin
+        day_night:
+            Fills {day_night} in expected string (Also changes Sunny to Clear)
+        model_id_attr:
+            Argument for WeatherSymbols. Triggers checking for this attribute on output cube.
+        record_run_attr:
+            Argument for WeatherSymbols. Triggers checking for this attribute on output cube.
+    """
+    result = WeatherSymbols(
+        wxtree=wxcode_decision_tree(),
+        model_id_attr=model_id_attr,
+        target_period=3600,
+        record_run_attr=record_run_attr,
+    )(wxcode_inputs)
+    if expected == "Masked":
+        assert result.data.mask
+    else:
+        assert WX_DICT[int(result.data)] == expected.format(
+            day_night=day_night
+        ).replace("Sunny_Night", "Clear_Night")
+
+    assert isinstance(result, iris.cube.Cube)
+    attributes = result.attributes.copy()
+    assert all(attributes.pop("weather_code") == list(WX_DICT.keys()))
+    assert attributes.pop("weather_code_meaning") == " ".join(WX_DICT.values())
+    assert attributes.pop("source") == "Unit test"
+    assert attributes.pop("institution") == "Met Office"
+    assert attributes.pop("title") == "Post-Processed IMPROVER unit test"
+    if model_id_attr:
+        assert attributes.pop(model_id_attr) == "uk_det uk_ens"
+    if record_run_attr:
+        assert (
+            attributes.pop(record_run_attr)
+            == "uk_det:20171109T2300Z:\nuk_ens:20171109T2100Z:"
+        )
+    assert not attributes
+
+    assert result.dtype == np.int32
+
+    for cube in wxcode_inputs:
+        if cube.coord("time").has_bounds():
+            for coord in ["time", "forecast_period"]:
+                assert result.coord(coord) == cube.coord(coord)
+            break
+
+
+@pytest.mark.parametrize("record_run_attr", (None, "mosg__model_run"))
+@pytest.mark.parametrize("model_id_attr", (None, "mosg__model_configuration"))
+@pytest.mark.parametrize("hour, day_night", ((12, "Day"), (0, "Night")))
+@pytest.mark.parametrize(
+    "hail_accum, hail_rate, lightning, rain, rain_vic, shower_condition, sleet, sleet_vic, "
+    "snow, snow_vic",
+    ((0, 0, 0, [0, 0, 0], [0, 0, 0], 0, [0, 0, 0], [0, 0, 0], [0, 0, 0], [0, 0, 0],),),
+)
+@pytest.mark.parametrize(
+    "expected, cloud, low_cloud, vis",
+    (
+        ("Sunny_{day_night}", [0, 0], 0, [0, 0]),
+        ("Partly_Cloudy_{day_night}", [1, 0], 0, [0, 0]),
+        ("Cloudy", [1, 1], 0, [0, 0]),
+        ("Overcast", [1, 1], 1, [0, 0]),
+        ("Mist", [1, 1], 1, [0, 1]),
+        ("Fog", [1, 1], 1, [1, 1]),
+    ),
+)
+def test_dry_routes(wxcode_inputs, day_night, model_id_attr, record_run_attr, expected):
```

Comment:
This test and the other `_routes` ones below it are running through many combinations of values which are set up to cover the branches of the tree.
There's a large amount of effort here on achieving test coverage (more like *tree* coverage than code coverage) - one or more manually written test cases for each leaf of the decision tree, demonstrating that it's possible to reach each of the leaves.
I may be missing some background and context, but to me this focus on tree coverage doesn't seem like good value for testing effort - kind of like it's unclear what the aims of testing are, so something has been done to be able to say "all tested, no bugs found".

If the goal is to check that all the tree leaves are reachable (or other properties of the configured tree), a white box testing approach could better separate the code/tree from the tests, "build the tests from the tree". For example, write code to start at the tree's leaves, traverse the tree towards the root to find split point values, then use those found values to test that the leaves are all reachable.

If the goal is to check that a variety of weather situations result in the expected symbols, it may be easier to express (and maintain) that test dataset as a big spreadsheet-style table, keep it as a CSV and then write one test function which processes the rows of the table. (IMPROVER already depends on pandas, so all that functionality is available).

If the goal is to check that the decision tree evaluation code correctly processes the tree as specified, that could be done with a smaller, specially crafted tree which exercises all the decision tree dict options as described in the documentation page.